pyproject.toml

```diff
@@ -70,6 +70,7 @@ matplotlib = "^3.7.0"
 qibojit = { git = "https://github.com/qiboteam/qibojit.git" }
 tensorflow = { version = "^2.14.1,<2.16", markers = "sys_platform == 'linux'" }
 qibotn = { git = "https://github.com/qiboteam/qibotn.git" }
+stim = "^1.12.0"
```

Comment:
`stim` is a very light dependency, but I'm becoming more and more cautious about adding new deps.

@renatomello it's fine to add it, but I'd like to split the `quantum_info` module soon (as soon as we'll have `qibo-core` available), if possible, to isolate its dependencies. Are you okay with that?

This is not related to this PR.

Reply:
what do you mean by split?

Reply:
Distributing in a separate package, like they are thinking for `qiboml`. It doesn't have to be a separate repo.

The problem is that, even if you want to just use the basic `NumpyBackend`, at the moment you're forced to get all the dependencies.

I proposed to have pip extras for each module, but after discussing it seems more intuitive to do:
```sh
pip install qibo qiboml qibo_info
```
than
```sh
pip install qibo[ml,info]
```
(I'm not sure myself how much pip extras are known to users, and I got convinced by the people involved in the discussion)

Reply:
I don't agree with that. The info module comes with very fundamental functions and should come with "raw" `qibo`.

This is the same in the other packages that also contain an info module (e.g. [Qiskit](https://github.com/Qiskit/qiskit/tree/f4b50fbb2d3fe7871589fae898f6eadbad7c959b/qiskit)) as well as in others where the same capabilities are spread out throughout their code.

Reply:
Well, then you should split `Clifford` out of `quantum_info` (or just the `stim` part).

Ideally, `qibo` should have mandatory dependencies only on `numpy` (`scipy` is also acceptable).
E.g., right now it is impossible to install Qibo on Mac with compiling a wheel with CMake and additional dependencies, because of `qdldl` (a transitive dependency of `cvxpy`). Moreover, even if you want to have a minimal execution, you will be downloading >200MB of TensorFlow, that is also preventing us from allowing Python 3.12.

Dependencies are an issue, and they should be carefully reviewed. That's why whatever add further dependencies should be optional, one way or the other.

Reply:
> Well, then you should split `Clifford` out of `quantum_info` (or just the `stim` part).
> 
> Ideally, `qibo` should have mandatory dependencies only on `numpy` (`scipy` is also acceptable). E.g., right now it is impossible to install Qibo on Mac with compiling a wheel with CMake and additional dependencies, because of `qdldl` (a transitive dependency of `cvxpy`). Moreover, even if you want to have a minimal execution, you will be downloading >200MB of TensorFlow, that is also preventing us from allowing Python 3.12.
> 
> Dependencies are an issue, and they should be carefully reviewed. That's why whatever add further dependencies should be optional, one way or the other.

- `stim` is an optional dependency. 
- `tensorflow` being in main `qibo` or not does not affect `quantum_info` because the module is backend-agnostic. 
- `cvxpy` is also an optional dependency.

Reply:
TensorFlow was a broader example, and it will in QiboML. I'm not concerned about `quantum_info` alone, the general problem are Qibo deps.

`stim` is currently an undeclared dependency, because you only added it to the test group, so it will never be installed by the user. `cvxpy` is an optional dependency with no associated extra, so it is once more useless for the end user (but it is flagged by Poetry anyhow when you are adding Qibo as a dep to another project).

> Right now `qibo` only depends on `stim` and `cvxpy` for tests, which is kind of inevitable with respect to the development side, but it's completely optional to an user.

Well, not really, since if you want to test the `NumpyBackend` you also need `cvxpy`. Splitting to a different package would decouple the two.

It could be in this repo as usual, just in `src/qibo_info` (or `src/quantum_info` as well, if that name is still available), just a different entry on PyPI.

> Moreover, if you have ideas to rewrite the convex optimization in the `diamond_norm` function without depending on `cvxpy`, I'm open to suggestions.

The goal is not to have a different implementation, that one is perfectly fine (I didn't review myself, but I'm not going to do it now). But features that require complex tools are not basic, even for end users (since they will need those tools).
I can also accept that extras are not intuitive, but requiring an additional package as a plugin should be fairly acceptable.

My concern is coming from the maintenance of the tools we are using, since Qibo is growing, and requiring many pieces to work all at the same time is more and more constraining.
Here is the status of `qdldl` maintenance:
https://github.com/osqp/qdldl/graphs/contributors
and here an illustrative cartoon about my concerns:
https://xkcd.com/2347/
:P

Reply:
> `stim` is currently an undeclared dependency, because you only added it to the test group, so it will never be installed by the user. 

As it should be. If the user calls `stim` without having it installed, the code explicitly raises an error telling the user to install `stim`. Also in the documentation is is clear the `stim` is a separate package that should be installed separately.

`cvxpy` is an optional dependency with no associated extra, so it is once more useless for the end user (but it is flagged by Poetry anyhow when you are adding Qibo as a dep to another project).

It could be an undeclared dependency too, it makes no difference to the implementation.

> > Right now `qibo` only depends on `stim` and `cvxpy` for tests, which is kind of inevitable with respect to the development side, but it's completely optional to an user.
> 
> Well, not really, since if you want to test the `NumpyBackend` you also need `cvxpy`. Splitting to a different package would decouple the two.

i.e. it only depends on `cvxpy` for test, like I mentioned before.
 
> It could be in this repo as usual, just in `src/qibo_info` (or `src/quantum_info` as well, if that name is still available), just a different entry on PyPI.

I already said my opinion on this. In any case, this is both a separate issue and something that more people should opine on it besides just the both of us.

Reply:
> I already said my opinion on this. In any case, this is both a separate issue and something that more people should opine on it besides just the both of us.

The discussion is openly ongoing both in #1117 and https://github.com/qiboteam/qibo-core